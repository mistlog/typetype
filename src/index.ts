export * from "./parser";
export * from "./generator";
export * from "./core";