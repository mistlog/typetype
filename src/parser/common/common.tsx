import { ReactPeg } from "react-peg";

// white space
export function _() {
    return (
        <repeat type="*">
            <set> \t\n\r</set>
        </repeat>
    );
}

export const Text = (content: string) => {
    return (
        <list>
            <_ /><text>{content}</text><_ />
        </list>
    )
}