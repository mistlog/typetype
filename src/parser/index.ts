export * from "./common";
export * from "./basic";
export * from "./function";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";