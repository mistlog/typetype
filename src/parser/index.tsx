export * from "./common";
export * from "./basic";
export * from "./expression";
export * from "./statement";
export * from "./declaration";
export * from "./start";