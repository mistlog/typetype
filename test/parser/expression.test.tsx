import { ReactPeg } from "react-peg";
import { UnionType, IntersectionType } from "../../src";
import { loadType, saveAST } from "../common";

describe("parser: expression", () => {
    test("UnionType", () => {
        const parser = ReactPeg.render(<UnionType />);
        const ast = parser.parse(`union ["0", "1"]`);
        saveAST(ast, "UnionType.json");
        expect(ast).toMatchSnapshot();

        {
            const ast = parser.parse(`| ["0", "1"]`);
            expect(ast).toMatchSnapshot();
        }
    })

    test("UnionType: special char", () => {
        const parser = ReactPeg.render(<UnionType />);
        const name = "UnionType-SpecialChar";
        const code = loadType(name);

        const ast = parser.parse(code);
        saveAST(ast, `${name}.json`);
        expect(ast).toMatchSnapshot();
    })

    test("IntersectionType", () => {
        const parser = ReactPeg.render(<IntersectionType />);
        const ast = parser.parse(`combine ["0", "1"]`);
        saveAST(ast, "IntersectionType.json");
        expect(ast).toMatchSnapshot();

        {
            const ast = parser.parse(`& ["0", "1"]`);
            expect(ast).toMatchSnapshot();
        }
    })
})