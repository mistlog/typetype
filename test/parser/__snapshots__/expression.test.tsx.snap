// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser: expression ContextType 1`] = `
Object {
  "body": Object {
    "context": "use js",
    "source": "const type = \\"1\\";
return type;

",
  },
  "kind": "ContextType",
}
`;

exports[`parser: expression ContextType: with resolver 1`] = `
Object {
  "kind": "StringTypeLiteral",
  "value": "resolved",
}
`;

exports[`parser: expression IntersectionType 1`] = `
Object {
  "kind": "IntersectionType",
  "types": Array [
    Object {
      "kind": "StringTypeLiteral",
      "value": "0",
    },
    Object {
      "kind": "StringTypeLiteral",
      "value": "1",
    },
  ],
}
`;

exports[`parser: expression IntersectionType 2`] = `
Object {
  "kind": "IntersectionType",
  "types": Array [
    Object {
      "kind": "StringTypeLiteral",
      "value": "0",
    },
    Object {
      "kind": "StringTypeLiteral",
      "value": "1",
    },
  ],
}
`;

exports[`parser: expression UnionType 1`] = `
Object {
  "kind": "UnionType",
  "types": Array [
    Object {
      "kind": "StringTypeLiteral",
      "value": "0",
    },
    Object {
      "kind": "StringTypeLiteral",
      "value": "1",
    },
  ],
}
`;

exports[`parser: expression UnionType 2`] = `
Object {
  "kind": "UnionType",
  "types": Array [
    Object {
      "kind": "StringTypeLiteral",
      "value": "0",
    },
    Object {
      "kind": "StringTypeLiteral",
      "value": "1",
    },
  ],
}
`;

exports[`parser: expression UnionType: special char 1`] = `
Object {
  "kind": "UnionType",
  "types": Array [
    Object {
      "kind": "StringTypeLiteral",
      "value": " ",
    },
    Object {
      "kind": "StringTypeLiteral",
      "value": "\\\\n",
    },
    Object {
      "kind": "StringTypeLiteral",
      "value": "\\\\t",
    },
  ],
}
`;
